### PR TITLE
create a basic useCdp hook using dai.js math fncs & refactor the generate sidebar to use it

### PR DESCRIPTION
**Pull request by jparklev**

### CODE REVIEW

**Comment by jparklev** [comment]
tests for the current production implementation of `generate.js` will be merged in here once ready

File: src/components/Sidebars/Generate.js [hunk 1]
```diff
@@ -1,84 +1,31 @@
...
   const { maker, newTxListener } = useMaker();
-  const [daiAvailable, setDaiAvailable] = useState(0);
-  const [liquidationPrice, setLiquidationPrice] = useState(0);
-  const [collateralizationRatio, setCollateralizationRatio] = useState(0);
+  const [amount, setAmount] = useState('');
 
-  const [storeState] = useStore();
-  const cdp = getCdp(cdpId, storeState);
-  const { liquidationRatio } = cdp;
+  const cdp = useCdp(cdpId);
```
Comment (michaelelliot) [inline src/components/Sidebars/Generate.js]: So much cleaner. :100:

File: src/components/Sidebars/Generate.js [hunk 2]
```diff
@@ -99,22 +46,13 @@ const Generate = ({ cdpId, reset }) => {
...
-          failureMessage={amountErrors}
-        />
-        <RatioDisplay
-          type={RatioDisplayTypes.CARD}
-          ratio={collateralizationRatio}
-          ilkLiqRatio={liquidationRatio}
-          text={lang.action_sidebar.generate_warning}
-          onlyWarnings={true}
-          show={amount !== '' && amount > 0 && !undercollateralized}
-          textAlign="center"
```
Comment (padraic-X) [inline src/components/Sidebars/Generate.js]: These warnings were part of CDP-533? Is there any reason for them to be removed?

File: src/components/Sidebars/Generate.js [hunk 3]
```diff
@@ -124,22 +62,20 @@ const Generate = ({ cdpId, reset }) => {
...
-              ratio={collateralizationRatio}
-              ilkLiqRatio={liquidationRatio}
-              text={formatCollateralizationRatio(collateralizationRatio)}
-              show={amount !== '' && amount > 0 && !undercollateralized}
-            />
+            <Text color={cdpUnderCollateralized ? 'orange.600' : null}>
+              {cdp
+                .collateralizationRatio({ dart: amountToGenerate })
+                .toString()}
+            </Text>
```
Comment (padraic-X) [inline src/components/Sidebars/Generate.js]: Just with the colours, I've added a [function](https://github.com/makerdao/mcd-cdp-portal/blob/8a13ff8a573450eb3a5fdf272e343eee90398a7c/src/styles/theme.js#L81) in themes.js that centralizes a colour scheme for several components that uses safety levels within the app. I feel like we should derive colours based on the underlying values where we can?

One thought I had looking at this PR is we could remove `ratio` and `ilkLiqRatio` as args from `RatioDisplay` and pass in the cdp object directly?

File: src/components/Sidebars/Generate.js [hunk 4]
```diff
@@ -1,84 +1,31 @@
...
-      liquidationPrice,
-      collateralizationRatio,
-      daiAvailable
-    } = calcCDPParams({
-      ilkData: cdp,
-      gemsToLock: collateralAmount,
-      daiToDraw: add(debtAmount, amountToGenerate)
-    });
+  const dustLimit = cdp.dust || 0;
+  const amountToGenerate = amount || 0;
```
Comment (padraic-X) [inline src/components/Sidebars/Generate.js]: Also going back on the discussion with BigNumbers, could these be `BigNumber(0)` instead?

File: src/components/Sidebars/Generate.js (hunk 2 above)
Comment (jparklev) [reply src/components/Sidebars/Generate.js]: nah, just got lost in a rebase. plan to go through and merge them back manually

File: src/components/Sidebars/Generate.js (hunk 3 above)
Comment (jparklev) [reply src/components/Sidebars/Generate.js]: >  I feel like we should derive colours based on the underlying values where we can?
👍👍

> One thought I had looking at this PR is we could remove ratio and ilkLiqRatio as args from RatioDisplay and pass in the cdp object directly?
ah yep. another option is to just pass the `cdp id` in and have `RatioDisplay` rely on a `useCdp` hook under the hood.

File: src/components/Sidebars/Generate.js (hunk 4 above)
Comment (jparklev) [reply src/components/Sidebars/Generate.js]: imo if we have `cdp.dust` coming in as a `bignumber` (which would be a reasonable change), then `BigNumber(0)` would make sense there. in the `amount` case, i think the raw value from input's `onChange` is one of the few places where it's ok not to use big numbers. we could at some point create a big number input component similar to [what 0x uses](https://github.com/0xProject/0x-launch-kit-frontend/blob/662daf332e9f09b498c17fa50ed84b9c37e1c990/src/components/common/big_number_input.tsx#L93) that uses the raw value internally, but casts it as a big number for the onChange function passed in as a prop.

File: src/components/Sidebars/Generate.js (hunk 3 above)
Comment (padraic-X) [reply src/components/Sidebars/Generate.js]: Perfect, better again!

File: src/components/Sidebars/Generate.js (hunk 4 above)
Comment (padraic-X) [reply src/components/Sidebars/Generate.js]: Ya, that's a great idea. I'd be on board with that

Comment (levity) [reply src/components/Sidebars/Generate.js]: If it's meant to represent the input value, it would be appropriate for it to be a string

**Comment by levity** [comment]
This is great, definitely a step in the right direction

File: src/components/Sidebars/Generate.js (hunk 4 above)
Comment (jparklev) [reply src/components/Sidebars/Generate.js]: agreed

**Comment by jparklev** [comment]
@b-pmcg in your view, does #178 supersede this PR?

**Comment by b-pmcg** [comment]
@jparklev yep, this one can probably be closed now.

**Pull request closed without merging**